### ISSUE TITLE
[REQUEST] Check that configuration is valid in a non-interactive way

### ISSUE THREAD

**Issue by lobre** [mawww/kakoune#4796]
### Feature

I think it would be valuable to have a simple way to check that the user configuration is valid, in a non-interactive manner. It would be checkable from a script, totally outside of kakoune.

I was expecting something like having a `-check` option that validates the configuration.

```
kak -check || echo "configuration is invalid"
```

As of now, I did not find a simple solution to do this. After having discussed on IRC, a suggestion was to dump the `*debug*` buffer into a file, and check if things were written in there. But it is rather cumbersome and only checks the number of lines, which is not super robust.

```
kak -e 'b *debug*; w debug; q'
if [[ "$(wc -l < debug)" -gt "1" ]]; then echo "configuration is invalid"; fi
```

Is there a better way? Should it be a feature of Kakoune?

I also try something like `kak -n -e 'try %{ source ~/.config/kak/kakrc } catch %{ quit 1 }'` but as `-n` is used, the system configuration is not loaded and `colorscheme` is not defined. So it results in:

```
1:1: 'source': 18:1: 'colorscheme': no such command
```

### Usecase

In home-manager (a NixOS tool that allows defining user configurations using Nix), we would like to validate that the generated Kakoune configuration is valid.

See the following PR for reference.

https://github.com/nix-community/home-manager/pull/3418

**Comment by mawww** [comment]
> I also try something like `kak -n -e 'try %{ source ~/.config/kak/kakrc } catch %{ quit 1 }'` but as `-n` is used, the system configuration is not loaded and `colorscheme` is not defined. So it results in:

I do not think passing `-n` is correct here, a user configuration can refer to options declared in plugins for example, so you would need to ensure those plugins are loaded and you need to go through the bundled kakrc for that (except if you bypass entirely and load plugins through another way).

**Comment by Screwtapello** [comment]
I can think of a couple of reasons this would be difficult:

- "loading" a Kakoune config file means executing arbitrary shell scripts, so there's no way to "check" a file without arbitrary side-effects
- the only source of truth for "which config files are loaded" is the code in the system-wide `kakrc` file, which also converts errors into `*debug*` warnings

One approach might be to separate the current system-wide `kakrc` file into two parts:

- `common.kak` that defines the `colorscheme` command and contains the code that generates the list of possible config files.
- `kakrc` that sources `common.kak`, then tries to load each possible config file in turn and logs errors to `*debug*`

Then you could add a third file, `kakrc-check` that sources `common.kak`, then tries to load each possible config file in turn and exits on the first error. Then checking Kakoune config would be `kak -n -e 'source "%val{runtime}/kakrc-check"'`.

That would be a lot of messing about for debatable benefit, however.

**Comment by lobre** [comment]
> I do not think passing -n is correct here, a user configuration can refer to options declared in plugins for example, so you would need to ensure those plugins are loaded and you need to go through the bundled kakrc for that (except if you bypass entirely and load plugins through another way).

I agree that this is not correct. To check the configuration, I do need to load the system kakrc, so that autoload and colorscheme are working properly. And if I load the system kakrc, it means that the user configuration file cannot be sourced manually (with `:source`), but I will have to use `${kak_config}/kakrc`.

As implemented in the issue https://github.com/mawww/kakoune/issues/3072, `KAKOUNE_CONFIG_DIR` environment variable can be used to modify this `${kak_config}` location.

But if I don’t source the file manually, I cannot `try`/`catch` and fail with `quit 1`, so I am not sure how to report to the parent script that the loading of the configuration failed.

If someone has an idea, please tell me.

As proposed by @Screwtapello, having the current system `kakrc` split to allow a more granular and manual loading could be a solution. But I do agree that it is a lot of changes for such a simple need.

Something rather dirty that I could do is still using `-n` but sourcing this system `kakrc` myself after having defined `KAKOUNE_CONFIG_DIR`. This way I can use `:source` and it seems to be doing what I want. I am not sure if I can consider this as being clean but at least it works...

```
cd $(mktemp -d)
echo "colorscheme wrongtheme"
export KAKOUNE_CONFIG_DIR=$(pwd)
kak -n -e 'try %{ source /usr/share/kak/kakrc; quit 0 } catch %{ quit 1 }'
echo $?

# results to 1
```

What do you think of this solution?

**Comment by lobre** [comment]
If no one else has other solutions, I think I am going to go with the above and call it a day. Maybe I would still like to get your input @mawww before closing. Let me know. Thanks!

**Comment by Screwtapello** [comment]
The one problem with that solution is that the systemwide `kakrc` deliberately turns plugin import errors into `*debug*` buffer warnings. If you add a broken plugin to your configuration, that script won't detect it.

It will detect problems in the user `kakrc` though, so if that's your primary concern, it's good enough.
